gobblin-modules/gobblin-metadata/src/main/java/gobblin/metadata/provider/MetadataAwareFileSystem.java

```diff
@@ -0,0 +1,131 @@
+   */
+  public static final String METADATA_PROVIDER_CLASS = "metadataProviderClass";
+  /**
+   * URI that is used to instantiate {@link DatasetAwareMetadataProvider}.
+   */
+  public static final String METADATA_PROVIDER_URI = "metadataProviderUri";
+
+  public static class Factory<S extends ScopeType<S>> extends FileSystemInstrumentationFactory<S> {
+    @Override
+    public FileSystem instrumentFileSystem(FileSystem fs, SharedResourcesBroker<S> broker,
+        ConfigView<S, FileSystemKey> config) {
+      Config metaConfig = config.getConfig();
+      String metadataProviderClass = ConfigUtils.getString(metaConfig, METADATA_PROVIDER_CLASS, "");
+      log.info("Metadata provider class is: " + metadataProviderClass);
+      String metadataProviderUri = ConfigUtils.getString(metaConfig, METADATA_PROVIDER_URI, "");
+      log.info("Metadata provider URI is: " + metadataProviderUri);
+      try {
+        DatasetAwareMetadataProvider metadataProvider = null;
+        if (!metadataProviderClass.isEmpty()) {
+          metadataProvider = (DatasetAwareMetadataProvider) GobblinConstructorUtils
+              .invokeLongestConstructor(Class.forName(metadataProviderClass), metadataProviderUri);
+        }
+        return new MetadataAwareFileSystem(fs, metadataProvider);
+      } catch (ReflectiveOperationException e) {
+        throw new RuntimeException("Failed to initialize metadata provider.", e);
+      }
+    }
+  }
+
+  private DatasetAwareMetadataProvider metadataProvider;
+
+  public MetadataAwareFileSystem(FileSystem underlying, DatasetAwareMetadataProvider provider) {
+    super(underlying);
+    this.metadataProvider = provider;
+  }
+
+  @Override
+  public boolean mkdirs(Path f, FsPermission permission)
+      throws IOException {
+    FsPermission realPermission = getPermAtPathFromMetadataIfPresent(permission, f);
+    return super.mkdirs(f, realPermission);
+  }
+
+  @Override
+  public FSDataOutputStream create(Path f, FsPermission permission, boolean overwrite, int bufferSize,
+      short replication, long blockSize, Progressable progress)
+      throws IOException {
+    FsPermission realPermission = getPermAtPathFromMetadataIfPresent(permission, f);
+    return super.create(f, realPermission, overwrite, bufferSize, replication, blockSize, progress);
+  }
+
+  @Override
+  public void setPermission(Path f, final FsPermission permission)
+      throws IOException {
+    super.setPermission(f, getPermAtPathFromMetadataIfPresent(permission, f));
+  }
+
+  @Override
+  public void setOwner(Path f, String user, String group)
+      throws IOException {
+    super.setOwner(f, user, getGroupAtPathFromMetadataIfPresent(group, f));
+  }
+
+  private String getGroupAtPathFromMetadataIfPresent(String defaultGroup, Path path) {
+    String realGroup = defaultGroup;
+    if (this.metadataProvider != null) {
+      realGroup = metadataProvider.getGlobalMetadataForDataset(path.toString()).getGroupOwner();
```

Comment:
Created a `DatasetAwareFsMetadataProvider` which can parse `datasetUrn` from a given path. Changed `MetadataAwareFileSystem` to use `DatasetAwareFsMetadataProvider` instead.